### DISCUSSION TITLE
Generate Generic Types to interact with models (for example models whereInput)

### DISCUSSION THREAD

**Discussion by SimoneLutero** [prisma/prisma#24911] [Q&A]
### Question

Hello, I've find a solution to infer a <model>WhereInput type knowing the model, but I would like to know if there is a smarter way or even better a method to generate a generic type.

### Expected behavior (optional)

For example if I have 2 models "User" and "Pippo" this is what I'm doing now and what I would like to be generated:

```

export type ModelWhereInput<Entity> =
  Entity extends User ? Prisma.UserWhereInput :
  Entity extends Pippo ? Prisma.PippoWhereInput :
  never

```
I can do this writing all by myself, but it costs so much and I have to manually edit this everytime schema changes.
This can be useful for generic methods in case your models has identic behavior in some fields.

**Comment by ludralph** [comment]
Hi @SimoneLutero 👋 

You can try creating a type map that includes all your models and their corresponding WhereInput types. This type map will help you infer the WhereInput type generically.
```ts
import { Prisma } from '@prisma/client';

// Define a type map for models
export type Models = {
  User: { 
    model: Prisma.User;
    where: Prisma.UserWhereInput;
    // other user related types
  },
  Pippo: {
    model: Prisma.Pippo;
    where: Prisma.PippoWhereInput;
    // other pippo related types
  }
  // Add other models here
};
```
Create a Generic Type to Infer WhereInput: Define a generic type that infers the WhereInput type based on the model key.

// Generic type to infer WhereInput
```ts
export type WhereInput<T extends keyof Models> = Models[T]['where'];
```

Use the Generic Type in Your Functions: You can now use this generic type in your functions to infer the WhereInput type for different models.

// Example usage
```ts
function can<TModel extends keyof Models>(action: string, subject: TModel, conditions: WhereInput<TModel>) {
  // Your implementation here
}
```
```ts
async function exampleUsage() {
  const userConditions: WhereInput<'User'> = { email: 'example@example.com' };
  const pippoConditions: WhereInput<'Pippo'> = { name: 'Pippo' };

  can('read', 'User', userConditions);
  can('update', 'Pippo', pippoConditions);
}

exampleUsage();
```
With this approach, you only need to update the Models type map.

**Reply by SimoneLutero** [reply]
Hi, thank you for the response. It's a good alternative solution, but I don't see a lot of difference with mine. Maybe I should try it as soon I will be able to and I will evaluate better.

Anyway it doesn't solve the problem of automatic generation.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
